modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/ConnectionClientPool.java

```diff
@@ -338,6 +359,43 @@ public GridCommunicationClient reserveClient(ClusterNode node, int connIdx) thro
         }
     }
 
+    private GridFutureAdapter<GridCommunicationClient> handleConnectExceptionWhileIncomingConnectionOccurred(
+            ConnectionKey connKey,
+            ConnectFuture origFut,
+            Throwable e
+    ) {
+        log.info("Outgoing connection failed while incoming connection attempt was detected. ");
+
+        GridFutureAdapter<GridCommunicationClient> triggerFut = new GridFutureAdapter<>();
+
+        triggerFut.listen(f -> {
```

Comment:
I'm not yet convinced that this code is thread-safe, so please add an explanation of why it should work. I see no tests that would prove it, so I have to trust your words.

Reply:
AFAIK, listen guarantees one time execution on future completion.
Future will be completed either on line 393 in this class or in InboundConnectionHandler:711.
Map clientFuts is Concurrent